egs/librispeech/asr/simple_v1/decode.py

```diff
@@ -0,0 +1,175 @@
+from lhotse.dataset import SpeechRecognitionDataset
+from lhotse.dataset.speech_recognition import K2DataLoader, K2SpeechRecognitionDataset, \
+    K2SpeechRecognitionIterableDataset, concat_cuts
+from lhotse.recipes.librispeech import download_and_untar, prepare_librispeech, dataset_parts_full
+
+from common import load_checkpoint
+from common import setup_logger
+from model import Model
+
+
+def decode(dataloader, model, subsampling, device, LG, symbols):
+    results = []  # a list of pair (ref_words, hyp_words)
+    for batch_idx, batch in enumerate(dataloader):
+        feature = batch['features']
+        supervisions = batch['supervisions']
+        supervision_segments = torch.stack(
+            (supervisions['sequence_idx'],
+             torch.floor_divide(supervisions['start_frame'], subsampling),
+             torch.floor_divide(supervisions['num_frames'], subsampling)),
+            1).to(torch.int32)
+        texts = supervisions['text']
+        assert feature.ndim == 3
+
+        feature = feature.to(device)
+        # at entry, feature is [N, T, C]
+        feature = feature.permute(0, 2, 1)  # now feature is [N, C, T]
+        with torch.no_grad():
+            nnet_output = model(feature)
+        # nnet_output is [N, C, T]
+        nnet_output = nnet_output.permute(0, 2,
+                                          1)  # now nnet_output is [N, T, C]
+
+        dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+        assert LG.is_cuda()
+        assert LG.device == device
+        assert nnet_output.device == device
+        # TODO(haowen): with a small `beam`, we may get empty `target_graph`,
+        # thus `tot_scores` will be `inf`. Definitely we need to handle this later.
+        lattices = k2.intersect_dense_pruned(LG, dense_fsa_vec, 2000.0, 20.0,
+                                             30, 300)
+        best_paths = k2.shortest_path(lattices)
+        best_paths = best_paths.to('cpu')
+        assert best_paths.shape[0] == len(texts)
+
+        for i in range(len(texts)):
+            hyp_words = [
+                symbols.get(x) for x in best_paths[i].aux_labels if x > 0
+            ]
+            results.append((texts[i].split(' '), hyp_words))
+
+        if batch_idx % 10 == 0:
+            logging.info('Processed batch {}/{} ({:.6f}%)'.format(
+                batch_idx, len(dataloader),
+                float(batch_idx) / len(dataloader) * 100))
+
+    return result
+
+
+def main():
+    # Caution(haowen): this script can not be run now, there are something
+    # wrong in determinize, we are still debugging the code
+    return
+    # load L, G, symbol_table
+    lang_dir = 'data/lang_nosp'
+    symbol_table = k2.SymbolTable.from_file(lang_dir + '/words.txt')
+
+    if not os.path.exists(lang_dir + '/LG.pt'):
+        print("Loading L_disambig.fst.txt")
+        with open(lang_dir + '/L_disambig.fst.txt') as f:
+            L = k2.Fsa.from_openfst(f.read(), acceptor=False)
```

Comment:
Could we have comments stating what's the input and output label when the FSTs are loaded (e.g. "L input symbols: phones; output symbols: words) and when they are inverted? I think it'd be easier to follow.

Reply:
Good idea, will try to add it at some point.  But remember, in this framework it's not really input and output symbols: it's symbols and aux_symbols.  I'd rather be consistent on this point.

Reply:
Got it. In that case, it would be super useful if you had a documentation page that presents a mapping between OpenFST concepts and K2 concepts (maybe you have it already).